### PR TITLE
Refactor event & properties for async server-side loading

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

This PR tackles the problem that came up with a customer who has more than 3MB worth of event names and properties, which significantly slows down PostHog. With this change, we'll now split all event and property definitions & metadata into their own models, enabling separate requests with pagination and server-side search.

To avoid creating another massive PR, this will be accomplished in two separate PRs. This PR **only affects the backend** (see details below), **except for the following pending backend change:**
- We don't yet remove `event_names`, `event_properties_with_usage`, ... and old related fields from the serializer response in `/api/projects/` endpoint. The reason for this is that this PR is then safe to merge without the next PR changing the frontend as all changes are incremental. (view 99c3b441af0f85ecb32696927712858d755e17d0)

## Details
- Introduces `EventDefintion` & `PropertyDefinition` models to store event & property information respectively (including usage, and whether each properly is numerical).
- Temporarily we're keeping the new models in sync with the values stored in `Team` using Django signals and an asynchronous task that adds, updates & removes any changed event names or properties. This syncing mechanism isn't particularly performance-optimized, but as this is a temporary measure (and async), it's not worth spending too much time on it right now.
- We add the `/api/projects/@current/event_definitions/` and `/api/projects/@current/property_definitions` endpoints to fetch event & properties definitions respectively.
    - The endpoint fully supports pagination to load items in batches.
    - The endpoint also supports server-side searching on event names. We'll have to do some field tests on the performance of this, as it might still be best to rely on frontend searching.
-  Bunch of relevant tests for points above.
- Adds deprecation notices across different points in the app in preparation for subsequent steps (see next steps below) [easily searchable with `TODO: #4070`] 
- Fixes some minor typing errors along the way.

## Next Steps
1. We need the matching PR that does the required changes in the frontend to use this new endpoint (plus add the pending backend change referenced above).
    - Particularly we need to implement pagination on these requests and load properties and events in batches to avoid server timeouts.
    - Depending on some field tests, we might want to prefer to do server-side querying altogether, or probably a combination of both while we can load all the events and properties to the frontend.
2. We need to update the plugin-server & related ingestion to start storing the event and properties directly in the new models.
3. We need to update the core backend to:
    - Compute event and property volume and usage directly in the new model.
    - Remove any deprecated code.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

File: posthog/models/event_definition.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from django.db import models
+
+from posthog.models.team import Team
+from posthog.models.utils import UUIDClassicModel
+
+
+class EventDefinition(UUIDClassicModel):
```
Comment (EDsCODE) [inline posthog/models/event_definition.py:7]: You're doing this with event properties too right?

Comment (paolodamico) [reply posthog/models/event_definition.py:7]: Yup, just wanted to do events first to get the heavy lifting out of the way

**Comment by mariusandra** [comment]
Hey! I went through the code and this seems solid. The logic to sync with the `team` model is probably fine for now, but I guess we'll remove it very soon. Mainly the syncing doesn't really sync right now, as the `team` almost always is updated from the plugin server, except if you toggle some flags on the project setting view (then it gets synced).

I did add one change to always have these as numbers (`0` not `null`), anticipating random effects downstream in the frontend...

image

**Review by mariusandra — APPROVED** [review]

File: posthog/models/event_definition.py (hunk 1 above)
Comment (Twixes) [reply posthog/models/event_definition.py]: Why `UUIDClassicModel` not `UUIDModel`?

**Pull request merged**